### PR TITLE
[WIP] Add synonyms and query rules support

### PR DESCRIPTION
**Pull request by CaiqueMitsuoka**
This PRs currently adds the support to read, search, create and update for synonyms and query rules.

It is based on the [Algolia Gem](https://github.com/algolia/algoliasearch-client-ruby) implementation and the [Algolia official documentation](https://www.algolia.com/doc/).

I would love to here your suggestions.

Ps: The query rule is a paid feature, and since I don't have a paid personal key I opened the PR in wip to run the test suite.

## Steps
- [x] Add support to export all the synonyms
- [x] Add support to save a batch of synonyms
- [x] Test both implementations
- [x] Add support to export all the query rules
- [x] Add support to save a batch of query rules
- [x] Test both implementations
- [ ] Add support to read a synonym(get by `objectID`)
- [ ] Add support to read a query rule(get by `objectID`)
- [ ] Add support to delete a synonym
- [ ] Add support to delete a query rule

### CODE REVIEW

File: lib/algolia.ex [hunk 1]
```diff
@@ -491,6 +491,149 @@ defmodule Algolia do
...
+        "nbHits" => 33,
+        "hits" => [%{
+          "objectID" => "1539816124475_0",
+          "synonyms" => ["big", "large"],
+          "type" => "synonym"
+          "_highlightResult" => %{...}
+        }, ...]
+      }}
+  """
+  def export_synonyms(index) do
```
Comment (britto) [inline lib/algolia.ex:513]: The official clients seem to have a iterative approach to these functions (like the Go client and the Scala client), which allows us to process one hit at a time. The Ruby one goes a step further and also returns the complete list at the end. I think we can achieve a similar flexibility in Elixir by leveraging Streams.

Another detail that I noticed in those clients is that they drop the `_highlightResult` key from every hit.

So I came up with this sketch of what could be a stream-based implementation:

```elixir
def export_synonyms(index) do
  get_all_paginated_hits(index, &search_synonyms/2)
end

def get_all_paginated_hits(index, search, hits_per_page \\ 100) do
  0
  |> Stream.iterate(&(&1 + 1))
  |> Stream.flat_map(fn page -> get_page_hits(index, search, page, hits_per_page) end)
  |> Stream.take_while(&(&1 != :stop))
end

defp get_page_hits(index, search, page, hits_per_page) do
  case search.(index, "", page: page, hits_per_page: hits_per_page) do
    {:ok, %{"hits" => hits, "nbPages" => pages}} when page + 1 < pages -> page_hits(hits)
    {:ok, %{"hits" => hits}} -> page_hits(hits) ++ [:stop]
    error -> [error, :stop]
  end
end

defp page_hits(hits) do
  Enum.map(hits, &{:ok, Map.drop(&1, ["_highlightResult"])})
end
```

This would return a `Stream` and thus allow users to consume it as they wish, including error handling. For example:

```elixir
"my_index"
|> Algolia.export_synonyms()
|> Enum.each(fn
  {:ok, %{"synonyms" => [from, to]}} -> IO.puts("#{from} and #{to} are the same")
  {:error, _} -> IO.puts("Unable to fetch all synonyms.")
end)
```

Or they can go ahead and just get the whole list, raising a match error if one of the page requests fails:

```elixir
"my_index"
|> Algolia.export_synonyms()
|> Enum.map(fn -> {:ok, hit} -> hit end)
```

How do you feel about that?

File: lib/algolia.ex [hunk 2]
```diff
@@ -491,6 +491,149 @@ defmodule Algolia do
...
+      |> Map.put("hitsPerPage", opts[:hits_per_page] || 20)
+      |> Map.drop([:page, :hits_per_page])
+      |> Jason.encode!()
+
+    :write
+    |> send_request(%{method: :post, path: Paths.search_synonyms(index), body: body})
+    |> inject_index_into_response(index)
+  end
+
+  @doc """
```
Comment (britto) [inline lib/algolia.ex:575]: Excellent! ❤️

File: test/algolia_test.exs [hunk 3]
```diff
@@ -4,19 +4,26 @@ defmodule AlgoliaTest do
   import Algolia
 
   @indexes [
-    "test", "test_1", "test_2", "test_3", "test_4",
```
Comment (britto) [inline test/algolia_test.exs]: I'd rather make a separate PR to `mix format` the whole repo instead, to make the review easier.

**Comment by davidbrusius** [comment]
@CaiqueMitsuoka is this ready for review? 😄
